### PR TITLE
Add limitrange mutation

### PR DESCRIPTION
**Pull request by y-kkamil**
**Description**

Changes proposed in this pull request:

- Adds possibility to create a limitrange.

### CODE REVIEW

File: components/console-backend-service/internal/gqlschema/schema.graphql [hunk 1]
```diff
@@ -930,6 +942,8 @@ type Mutation {
     deleteService(name: String!, namespace: String!): Service @HasAccess(attributes: {resource: "services", verb: "delete", apiGroup: "", apiVersion: "v1", namespaceArg: "namespace", nameArg: "name"})
 
     createNamespace(name: String!, labels: Labels): NamespaceCreationOutput! @HasAccess(attributes: {resource: "namespaces", verb: "create", apiGroup: "", apiVersion: "v1"})
+
+    createLimitRange(namespace: String!, name: String!, limits: LimitRangeInput!): LimitRange @HasAccess(attributes: {resource: "limitrange", verb: "create", apiGroup: "", apiVersion: "v1"})
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/gqlschema/schema.graphql]: Please change name of third parameter to `limitRange`.

File: components/console-backend-service/internal/domain/k8s/limitrange_resolver.go [hunk 2]
```diff
@@ -31,3 +31,9 @@ func (lr *limitRangeResolver) LimitRangesQuery(ctx context.Context, namespace st
 
 	return lr.converter.ToGQLs(limitRange), nil
 }
+
+func (lr *limitRangeResolver) CreateLimitRange(ctx context.Context, namespace string, name string, limitRange gqlschema.LimitRangeInput) (*gqlschema.LimitRange, error) {
+	createdLimitRange, err := lr.lister.Create(namespace, name, limitRange)
+
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_resolver.go]:
``` go
	if err != nil {
		glog.Error(errors.Wrapf(err, "while creating %s `%s` in namespace `%s`", pretty.LimitRange, name, namespace))
		return nil, gqlerror.New(err, pretty.LimitRange, gqlerror.WithName(name), gqlerror.WithNamespace(namespace))
	}
```

If we haven't pretty for LimitRange, please add it to relevant file, and use it above.

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go [hunk 3]
```diff
@@ -33,3 +39,54 @@ func (svc *limitRangeService) List(ns string) ([]*v1.LimitRange, error) {
 
 	return result, nil
 }
+
+func (svc *limitRangeService) Create(namespace string, name string, limit gqlschema.LimitRangeInput) (*v1.LimitRange, error) {
+	parsedMaxMemory, errMaxMemory := resource.ParseQuantity(*limit.Max.Memory)
+	if errMaxMemory != nil {
+		return nil, errors.Wrap(errMaxMemory, "Error while creating limit range, max.memory")
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service.go]:
``` go
return nil, errors.Wrapf(errMaxMemory, "Error while creating %s, max.memory", pretty.LimitRange)
```

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go [hunk 4]
```diff
@@ -33,3 +39,54 @@ func (svc *limitRangeService) List(ns string) ([]*v1.LimitRange, error) {
...
+	}
+
+	parsedDefaultMemory, errDefaultMemory := resource.ParseQuantity(*limit.Default.Memory)
+	if errDefaultMemory != nil {
+		return nil, errors.Wrap(errDefaultMemory, "Error while creating limit range, default.memory")
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service.go]:
``` go
return nil, errors.Wrapf(errMaxMemory, "Error while creating %s, default.memory", pretty.LimitRange)
```

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go [hunk 5]
```diff
@@ -33,3 +39,54 @@ func (svc *limitRangeService) List(ns string) ([]*v1.LimitRange, error) {
...
+	}
+
+	parsedDefaultRequestMemory, errDefaultRequestMemory := resource.ParseQuantity(*limit.DefaultRequest.Memory)
+	if errDefaultRequestMemory != nil {
+		return nil, errors.Wrap(errDefaultRequestMemory, "Error while creating limit range, defaultRequest.memory")
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service.go]:
``` go
return nil, errors.Wrapf(errMaxMemory, "Error while creating %s, defaultRequest.memory", pretty.LimitRange)
```

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go [hunk 6]
```diff
@@ -33,3 +39,54 @@ func (svc *limitRangeService) List(ns string) ([]*v1.LimitRange, error) {
...
+	}
+
+	if limit.Type != "Container" && limit.Type != "Pod" {
+		return nil, errors.New("Error while creating limit range, Type has to be one of: 'Container', 'Pod'")
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service.go]:
```go
return nil, errors.Wrapf(errMaxMemory, "Error while creating %s, Type has to be one of: 'Container', 'Pod'", pretty.LimitRange)
```

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go [hunk 7]
```diff
@@ -33,3 +39,54 @@ func (svc *limitRangeService) List(ns string) ([]*v1.LimitRange, error) {
...
+					DefaultRequest: v1.ResourceList{
+						v1.ResourceMemory: parsedDefaultRequestMemory,
+					},
+					Type: v1.LimitType(limit.Type),
+				},
+			},
+		},
+	}
+
+	result, err := svc.client.LimitRanges(namespace).Create(&limitRange)
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service.go]:
```go
return svc.client.LimitRanges(namespace).Create(&limitRange)
```

File: components/console-backend-service/internal/domain/k8s/limitrange_service_test.go [hunk 8]
```diff
@@ -48,3 +50,46 @@ func fixLimitRangeInformer(objects ...runtime.Object) cache.SharedIndexInformer
...
+	})
+
+	t.Run("Limit Range creation failed, wrong limit range type", func(t *testing.T) {
+		limitRangeGQL := createQueryFromProperties("512Mi", "512Mi", "512Mi", "RANDOM")
+		_, err := svc.Create(namespace, name, limitRangeGQL)
+		require.Error(t, err)
+	})
+}
+
+func createQueryFromProperties(defaultMem string, defaultRequestMem string, maxMem string, lrType string) gqlschema.LimitRangeInput {
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service_test.go]:
```go
func fixLimitRange(...) { ... }
```

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go (hunk 3 above)
Comment (sjanota) [reply components/console-backend-service/internal/domain/k8s/limitrange_service.go]: You may want to aggregate errors as @akucharska did in [resourcequota_service.go#L56](https://github.com/kyma-project/kyma/blob/master/components/console-backend-service/internal/domain/k8s/resourcequota_service.go#L56)

File: components/console-backend-service/internal/gqlschema/schema.graphql [hunk 9]
```diff
@@ -406,6 +406,18 @@ type LimitRangeItem {
...
 }
 
+input LimitRangeInput {
+    default: ResourceTypeInput!,
+    defaultRequest: ResourceTypeInput!
+    max: ResourceTypeInput!
+    type: String!
+}
+
+input ResourceTypeInput {
```
Comment (sjanota) [inline components/console-backend-service/internal/gqlschema/schema.graphql]: There is already a type for that: [`ResourceValuesInput `](https://github.com/akucharska/kyma/blob/eedde0c38f6ef7efe96f916592684681511c1533/components/console-backend-service/internal/gqlschema/schema.graphql#L583)

File: components/console-backend-service/internal/domain/k8s/limitrange_resolver_test.go [hunk 10]
```diff
@@ -14,7 +14,7 @@ func TestLimitRangeResolver_LimitRangeQuery(t *testing.T) {
 	// GIVEN
 	informer := fixLimitRangeInformer(fixLimitRange())
 
-	svc := newLimitRangeService(informer)
+	svc := newLimitRangeService(informer, nil)
 	testingUtils.WaitForInformerStartAtMost(t, time.Second, informer)
 
 	resolver := newLimitRangeResolver(svc)
```
Comment (sjanota) [inline components/console-backend-service/internal/domain/k8s/limitrange_resolver_test.go]: You're missing a test for creating LimitRange

Comment (y-kkamil) [reply components/console-backend-service/internal/domain/k8s/limitrange_resolver_test.go]: added

File: components/console-backend-service/internal/gqlschema/schema.graphql (hunk 9 above)
Comment (y-kkamil) [reply components/console-backend-service/internal/gqlschema/schema.graphql]: `ResourceTypeInput` removed, changed to `ResourceValuesInput`

File: components/console-backend-service/internal/gqlschema/schema.graphql (hunk 1 above)
Comment (y-kkamil) [reply components/console-backend-service/internal/gqlschema/schema.graphql]: renamed

File: components/console-backend-service/internal/domain/k8s/limitrange_resolver.go (hunk 2 above)
Comment (y-kkamil) [reply components/console-backend-service/internal/domain/k8s/limitrange_resolver.go]: done

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go (hunk 3 above)
Comment (y-kkamil) [reply components/console-backend-service/internal/domain/k8s/limitrange_service.go]: done

File: components/console-backend-service/internal/domain/k8s/resourcequota_service.go [hunk 11]
```diff
@@ -126,9 +126,9 @@ func (svc *resourceQuotaService) ListStatefulSets(namespace string) ([]*apps.Sta
 	return result, nil
 }
 
-func (svc *resourceQuotaService) ListPods(namespace string, labelSelector map[string]string) ([]v1.Pod, error) {
+func (svc *resourceQuotaService) ListPods(namespace string, envLabelSelector map[string]string) ([]v1.Pod, error) {
```
Comment (sjanota) [inline components/console-backend-service/internal/domain/k8s/resourcequota_service.go]: It's just labelSelector, it is not related any env label

File: components/console-backend-service/internal/domain/k8s/limitrange_service.go [hunk 12]
```diff
@@ -33,3 +43,55 @@ func (svc *limitRangeService) List(ns string) ([]*v1.LimitRange, error) {
 
 	return result, nil
 }
+
+func (svc *limitRangeService) Create(namespace string, name string, limit gqlschema.LimitRangeInput) (*v1.LimitRange, error) {
+	var errs apierror.ErrorFieldAggregate
+
+	parsedMaxMemory, errMaxMemory := resource.ParseQuantity(*limit.Max.Memory)
```
Comment (magicmatatjahu) [inline components/console-backend-service/internal/domain/k8s/limitrange_service.go]: Maybe extract logic for errors to separate function? This is only a proposition.

File: components/console-backend-service/internal/domain/k8s/resourcequota_service.go (hunk 11 above)
Comment (y-kkamil) [reply components/console-backend-service/internal/domain/k8s/resourcequota_service.go]: renamed

**Pull request merged**